### PR TITLE
Implements opinionated decorator lifecycle

### PR DESCRIPTION
**Pull request by elijahbenizzy**
This adds to the set of decorators that can be applied to a node,
and puts them in a uniform setting. The ordering is provided in
function_base.resolve_nodes -- the algorithm is as follows:

1. If there is a list of function resolvers, apply them one
after the other. Otherwise, apply the default function resolver
which will always return just the function. This determines whether to
proceed -- if any function resolver is none, short circuit and return
an empty list of nodes.

2. If there is a list of node creators, that list must be of length 1
-- this is determined in the node creator class. Apply that to get
the initial node.

3. If there is a list of node expanders, apply that. This must be a
list of length one. This gives out a list of nodes.

4. If there is a node transformer, apply that. Note that the node transformer
gets applied individually to just the sink nodes in the subdag. It subclasses
"DagTransformer" to do so.

5. Return the final list of nodes.

## Additions

- Transformer node

## Removals

-

## Changes

- Refactors decorator lifecycle to be clearer/more flexible

## Testing

1.

## Screenshots
If applicable

## Notes

-

## Todos

-

## Checklist

- [x] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [ ] Code follows the standards laid out in the [TODO link to standards]()
- [x] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python

- [ ] python 3.6
- [ ] python 3.7

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
TODO:
- [ ] Change to have `transform` be a superclass (much like `config`), with `to` as the method
- [ ] Add `transform` to documentation

File: hamilton/function_modifiers_base.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        :param nodes: Collection of nodes (not necessarily connected) to modify
+        :return: the new DAG of nodes
+        """
+        pass
+
+
+class NodeExpander(SubDAGModifier, NodeTransformLifecycle):
+    EXPAND_NODES = 'expand_nodes'
+
+    def transform_dag(self, nodes: Collection[node.Node], config: Dict[str, Any]) -> Collection[node.Node]:
```
Comment (skrawcz) [inline hamilton/function_modifiers_base.py:133]: doc string

File: hamilton/function_modifiers_base.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+
+class NodeTransformer(SubDAGModifier):
+    TRANSFORM_NODES = 'transform_nodes'
+
+    def _separate_final_nodes(self, nodes: Collection[node.Node]) -> Tuple[Collection[node.Node], Collection[node.Node]]:
+        """Separates out final nodes (sinks) from the nodes.
+
+        :param nodes: Nodes to separate out
+        :return: A list of nodes
```
Comment (skrawcz) [inline hamilton/function_modifiers_base.py:171]: doc string is incorrect

File: tests/resources/layered_decorators.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    return sum(kwargs.values())
+
+
+@config.when(foo='bar')
+@does(_sum)
+@parametrized(parameter='a', assigned_output={('e', 'First value'): 10, ('f', 'First value'): 20})
+@transform_to("c=c**2+d")
+def c(a: int, b: int) -> int:
+    """Demonstrates utilizing a bunch of decorators
+
```
Comment (skrawcz) [inline tests/resources/layered_decorators.py:14]: I think you need to explain how many "columns"/"nodes" this produces in the end.

**Review by skrawcz — COMMENTED** [review]
So what's net new here? Just the transform decorator? 

Otherwise the decorator lifecycle is largely a refactor of existing behavior, right?

**Comment by elijahbenizzy** [comment]
> So what's net new here? Just the transform decorator? Suggestion on name: `augment`. Transform is a bit overloaded a term.
> 
> Otherwise the decorator lifecycle is largely a refactor of existing behavior, right?

The decorator lifecycle is solidified and expanded -- moved some things around, but yeah, a refactor of existing behavior. Technically this allows models + parametrization together (whereas before it wouldn't), although that could get a little weird and probably won't be used much.

**Comment by elijahbenizzy** [comment]
TODO left prior to publishing:
- [ ] Fix @skrawcz's requests
- [ ] Add instructions for writing your own node lifecycle components
- [ ] Add tests to ensure that `@config.when` decorator works

File: tests/resources/layered_decorators.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+@config.when(foo='baz')
+def c__foobaz(a: int, b: int) -> int:
+    """Demonstrates utilizing a bunch of decorators.
+    In all, this outputs two total nodes.
+    - config.when makes it only apply when foo=bar
+    - @does makes it do the sum pattern
+    - @parametrized curries the function then turns it into two
+    - @augment changes the function to make each of those two final nodes take in vars c and d
+    """
+    pass
```
Comment (skrawcz) [inline tests/resources/layered_decorators.py:35]: err - do we actually want to support `augment`?

Reasoning:
- it is not clear at all what the outputs are here. 
- I would not want to debug this function
- what does foo refer to?

Comment (skrawcz) [reply tests/resources/layered_decorators.py:35]: yeah let's maybe hold on this to get feedback from others?

Comment (elijahbenizzy) [reply tests/resources/layered_decorators.py:35]: I like it but not necessary. It loses type information. Foo is just a config parameter, meant to demonstrate that multiple decorators work together.

File: hamilton/function_modifiers_base.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+                raise ValueError(f'Got multiple decorators for decorator @{self.__class__}. Only one allowed.')
+            curr_value = getattr(fn, lifecycle_name)
+            setattr(fn, lifecycle_name, curr_value + [self])
+        else:
+            setattr(fn, lifecycle_name, [self])
+        return fn
+
+
+class NodeResolver(NodeTransformLifecycle):
+    """Decorator to resolve a nodes function. Can modify anything about the function and is run before the node."""
```
Comment (skrawcz) [inline hamilton/function_modifiers_base.py:70]: is this more apt?
```suggestion
    """Decorator to resolve a nodes function. Can modify anything about the function and is run at DAG creation time."""
```

File: hamilton/function_modifiers_base.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        """Modifies a DAG consisting of a set of nodes. Note that this is to support the following two base classes.
+
+        :param nodes: Collection of nodes (not necessarily connected) to modify
+        :param config: Configuration in case any is needed
+        :return: the new DAG of nodes
+        """
+        pass
+
+
+class NodeExpander(SubDAGModifier):
```
Comment (skrawcz) [inline hamilton/function_modifiers_base.py:144]: doc string

File: hamilton/function_modifiers_base.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    EXPAND_NODES = 'expand_nodes'
+
+    def transform_dag(self, nodes: Collection[node.Node], config: Dict[str, Any], fn: Callable) -> Collection[node.Node]:
```
Comment (skrawcz) [inline hamilton/function_modifiers_base.py:147]: doc?

File: hamilton/graph.py [hunk 8]
```diff
@@ -23,6 +24,25 @@ def is_submodule(child: ModuleType, parent: ModuleType):
     return parent.__name__ in child.__name__
 
 
+def custom_subclass_check(requested_type: Type[Type], param_type: Type[Type]):
+    """This is a custom check as subclass does not work with python generic.
+    That said, its
```
Comment (skrawcz) [inline hamilton/graph.py:29]: unfinished comment.

File: tests/resources/layered_decorators.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+@does(_sum)
+@parametrized(parameter='a', assigned_output={('e', 'First value'): 10, ('f', 'First value'): 20})
+@config.when(foo='bar')
+def c__foobar(a: int, b: int) -> int:
+    """Demonstrates utilizing a bunch of decorators.
+    In all, this outputs two total nodes.
+    - config.when makes it only apply when foo=bar
+    - @does makes it do the sum pattern
+    - @parametrized curries the function then turns it into two
+    - @augment changes the function to make each of those two final nodes take in vars c and d
```
Comment (skrawcz) [inline tests/resources/layered_decorators.py:16]: remove

File: tests/resources/layered_decorators.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+@does(_sum)
+@parametrized(parameter='a', assigned_output={('e', 'First value'): 10, ('f', 'First value'): 20})
+@config.when(foo='bar')
+def c__foobar(a: int, b: int) -> int:
+    """Demonstrates utilizing a bunch of decorators.
+    In all, this outputs two total nodes.
+    - config.when makes it only apply when foo=bar
+    - @does makes it do the sum pattern
+    - @parametrized curries the function then turns it into two
```
Comment (skrawcz) [inline tests/resources/layered_decorators.py:15]: say what the names of the two nodes will be?
```suggestion
    - @parametrized curries the function then turns it into two; one named `e` and the other `f`.
```

File: tests/resources/layered_decorators.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+@does(_sum)
+@parametrized(parameter='a', assigned_output={('e', 'First value'): 11, ('f', 'First value'): 22})
+@config.when(foo='baz')
+def c__foobaz(a: int, b: int) -> int:
+    """Demonstrates utilizing a bunch of decorators.
+    In all, this outputs two total nodes.
+    - config.when makes it only apply when foo=bar
+    - @does makes it do the sum pattern
+    - @parametrized curries the function then turns it into two
+    - @augment changes the function to make each of those two final nodes take in vars c and d
```
Comment (skrawcz) [inline tests/resources/layered_decorators.py:30]: remove

File: tests/resources/layered_decorators.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+@does(_sum)
+@parametrized(parameter='a', assigned_output={('e', 'First value'): 11, ('f', 'First value'): 22})
+@config.when(foo='baz')
+def c__foobaz(a: int, b: int) -> int:
+    """Demonstrates utilizing a bunch of decorators.
+    In all, this outputs two total nodes.
+    - config.when makes it only apply when foo=bar
+    - @does makes it do the sum pattern
+    - @parametrized curries the function then turns it into two
```
Comment (skrawcz) [inline tests/resources/layered_decorators.py:29]: what is the names here?

File: tests/resources/layered_decorators.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    - @augment changes the function to make each of those two final nodes take in vars c and d
+    """
```
Comment (skrawcz) [inline tests/resources/layered_decorators.py:31]: probably worth commenting about the relationship between these two examples. They're both prefixed with `c`... I'm actually still confused what the output node names will be, given that e & f overlap...

Comment (elijahbenizzy) [reply tests/resources/layered_decorators.py:31]: They are mutually disjoint -- e and f will have different values depending on the value of `foo`

File: hamilton/function_modifiers.py [hunk 14]
```diff
@@ -106,8 +47,23 @@ def validate(self, fn: Callable):
             raise InvalidDecoratorException(
                 f'Annotation is invalid -- no such parameter {self.parameter} in function {fn}')
 
+    def expand_node(self, node_: node.Node, config: Dict[str, Any], fn: Callable) -> Collection[node.Node]:
```
Comment (skrawcz) [inline hamilton/function_modifiers.py:50]: I assume this is unit tested? :)

File: tests/resources/layered_decorators.py (hunk 13 above)
Comment (skrawcz) [reply tests/resources/layered_decorators.py:31]: Yeah, duh. But wasn't obvious on first pass :)

File: hamilton/function_modifiers.py (hunk 14 above)
Comment (elijahbenizzy) [reply hamilton/function_modifiers.py:50]: Yeah, it was before I even made this change. See `test_function_modifiers.py`

**Review by skrawcz — APPROVED** [review]

File: tests/resources/layered_decorators.py (hunk 13 above)
Comment (elijahbenizzy) [reply tests/resources/layered_decorators.py:31]: Added docs to clarify :)

File: tests/resources/layered_decorators.py (hunk 12 above)
Comment (elijahbenizzy) [reply tests/resources/layered_decorators.py:34]: E and f I think? Added docs.

**Pull request merged**